src/basic/path-util.c

```diff
@@ -794,6 +794,89 @@ char* dirname_malloc(const char *path) {
         return dir2;
 }
 
+static const char *skip_slash_or_dot(const char *p) {
+        for (; !isempty(p); p++) {
+                if (*p == '/')
+                        continue;
+                if (startswith(p, "./")) {
+                        p++;
+                        continue;
+                }
+                break;
+        }
+        return p;
+}
+
+int path_find_first_component(const char **p, bool accept_dot_dot, const char **ret) {
+        const char *q, *first, *end_first, *next;
+        size_t len;
+
+        assert(p);
+
+        /* When a path is input, then returns the pointer to the first component and its length, and
+         * move the input pointer to the next component or nul.
```

Comment:
maybe also add a sentence:

```
This skips both over any `/` immediately *before* and *after* the first component before returning.
```

I think that's important to realize when using the function.